### PR TITLE
[Kernel] Add API on TxnBuilder to set the table properties

### PR DESCRIPTION
**Pull request by EstherBear**
Add API on `TransactionBuilder` to set the table properties.

#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description

Adds an API to `TransactionBuilder` to set the table properties to provide a way to configure the table by committing a transaction. For example, user can enable inCommitTimestamp property with this API.

## How was this patch tested?

Adds unit tests when setting valid and invalid properties with this API.

## Does this PR introduce _any_ user-facing changes?

Yes, user can use this API to set table properties.

### CODE REVIEW

File: kernel/kernel-api/src/main/java/io/delta/kernel/TransactionBuilder.java [hunk 1]
```diff
@@ -69,6 +70,19 @@ TransactionBuilder withTransactionId(
             String applicationId,
             long transactionVersion);
 
+    /**
+     * Set the table properties for the table.
+     *
+     * @param engine     {@link Engine} instance to use.
+     * @param properties The table properties to set. These are key-value pairs that can be used to
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/TransactionBuilder.java:77]: specify the behavior when the table already contains the property with same key. i.e whether it gets replaced or not.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java [hunk 2]
```diff
@@ -65,6 +67,12 @@ public class TableConfig<T> {
     private final Function<String, T> fromString;
     private final Predicate<T> validator;
     private final String helpMessage;
+    private static final HashMap<String, TableConfig> entries = new HashMap<>();
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java:70]: nit: move the static variables to the top

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java [hunk 3]
```diff
@@ -86,13 +94,51 @@ private TableConfig(
...
+    public String getKey() {
+        return key;
+    }
+
+    /**
+     * Validates the given properties.
+     *
+     * @param configurations the properties to validate
+     *
+     * @throws IllegalArgumentException if any of the properties are invalid
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java:116]: Looks like we are also throwing `UnknownConfigurationKeyException`

nit: remove the new line between `@param` and `@throws` block.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java [hunk 4]
```diff
@@ -86,13 +94,51 @@ private TableConfig(
...
+     */
+    public static void validateProperties(Map<String, String> configurations) {
+        for (Map.Entry<String, String> kv : configurations.entrySet()) {
+            String key = kv.getKey();
+            String value = kv.getValue();
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java:121]: how does delta-spark deal with case sensitivity of the keys?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java [hunk 5]
```diff
@@ -86,13 +94,51 @@ private TableConfig(
...
+                }
+            } else {
+                throw DeltaErrors.unknownConfigurationKeyException(key);
+            }
+        }
+    }
+
+    private void validate(String value) {
+        T parsedValue = fromString.apply(value);
+        if (!validator.test(parsedValue)) {
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java:137]: Lets make this `IllegalArgumentException`, also a KernelException.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java [hunk 6]
```diff
@@ -85,6 +86,12 @@ public TransactionBuilder withTransactionId(
         return this;
     }
 
+    @Override
+    public TransactionBuilder withTableProperties(Engine engine, Map<String, String> properties) {
+        this.tableProperties = Optional.of(properties);
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java:91]: Do we need to make a copy of the properties? Use `Collections.unmodifiableMap` if needed

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java [hunk 7]
```diff
@@ -104,6 +111,14 @@ public Transaction build(Engine engine) {
         boolean isNewTable = snapshot.getVersion(engine) < 0;
         validate(engine, snapshot, isNewTable);
 
+        Metadata  metadata = snapshot.getMetadata().clone();
+        if (tableProperties.isPresent()) {
+            TableConfig.validateProperties(tableProperties.get());
+            Map<String, String> newConfiguration = new HashMap<>(metadata.getConfiguration());
+            newConfiguration.putAll(tableProperties.get());
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java:118]: Shouldn't we checking if the property already exists and can it be replaced? What does the Delta-Spark do?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java [hunk 8]
```diff
@@ -104,6 +111,14 @@ public Transaction build(Engine engine) {
...
+            metadata.setConfiguration(newConfiguration);
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java:119]: Add new method `Metadata.withNewConfiguration(Map<String, String> config)` that returns a copy of the `MEtadata` with new configuration. It also avoids the need for `MEtadata.clone`

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java [hunk 9]
```diff
@@ -151,9 +154,12 @@ private TransactionCommitResult doCommit(
             throws FileAlreadyExistsException {
         List<Row> metadataActions = new ArrayList<>();
         metadataActions.add(createCommitInfoSingleAction(generateCommitAction()));
+        if (shouldUpdateMetadata || isNewTable) {
+            metadataActions.add(createMetadataSingleAction(metadata.toRow()));
+            shouldUpdateMetadata = false;
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java:159]: why does this need to be reset?

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 10]
```diff
@@ -166,6 +167,59 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
...
+  test("create table - invalid properties - expect failure") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val txnBuilder = table.createTransactionBuilder(engine, testEngineInfo, CREATE_TABLE)
+
+      val ex1 = intercept[UnknownConfigurationKeyException] {
+        txnBuilder
+          .withSchema(engine, testSchema)
+          .withTableProperties(engine, Map("invalid key" -> "10").asJava)
+          .build(engine)
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:205]: make an utility method for this code block and use it it across tests. Simplifies the test code.

something like `Transaction createTxn(engine, testSchema, properties)`

Comment (vkorukanti) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:205]: If there is already an existing method, add the param for config.

**Review by vkorukanti — CHANGES_REQUESTED** [review]
Looks great! Thanks for making this PR.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java (hunk 6 above)
Comment (EstherBear) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java:91]: I think we don't need to modify the properties here? We just need to access it. So we don't need to make a copy here?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java (hunk 7 above)
Comment (EstherBear) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java:118]: I think[ Delta-Spark](https://github.com/delta-io/delta/blob/fa7d6c0eb5de0e05b0efac6b11cf96c58740a828/spark/src/main/scala/org/apache/spark/sql/delta/commands/alterDeltaTableCommands.scala#L157) replaces it for most of the cases.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java (hunk 9 above)
Comment (EstherBear) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java:159]: Because we provide a way to set the table properties which means we need to reset the metadata even when it's not a new table.

File: kernel/kernel-api/src/main/java/io/delta/kernel/TransactionBuilder.java (hunk 1 above)
Comment (dhruvarya-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/TransactionBuilder.java:78]: Also, we should ideally not update the metadata if the key-value pair matches exactly with what is there in the table.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java [hunk 11]
```diff
@@ -86,13 +97,48 @@ private TableConfig(
...
+     * Returns the key of the table property.
+     *
+     * @return the key of the table property
+     */
+    public String getKey() {
+        return key;
+    }
+
+    /**
+     * Validates the given properties.
```
Comment (dhruvarya-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java:115]: Please specify what exactly is being validated e.g. validates that the provided properties have the delta prefix in the key name ...

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 12]
```diff
@@ -166,6 +167,49 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
     }
   }
 
+  test("create table and set properties") {
```
Comment (dhruvarya-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:170]: Can you add one more test case here:
1. First commit is written with valid table properties
2. The second commit does not specify any table properties ---> the original properties should be retained.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java [hunk 13]
```diff
@@ -60,12 +65,18 @@ public class TableConfig<T> {
             "needs to be a positive integer."
     );
 
+    private static final HashMap<String, TableConfig> entries = new HashMap<>();
```
Comment (dhruvarya-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java:68]: What is entries? The set of all legal properties? Please use a more descriptive name and add a docstring as well.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java (hunk 9 above)
Comment (dhruvarya-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java:159]: I believe the rationale behind adding `shouldUpdateMetadata` is clear but what is unclear is why it needs to reset to `false`? We don't do that for `isNewTable`

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java [hunk 14]
```diff
@@ -86,13 +97,48 @@ private TableConfig(
...
+            }
         }
-        return value;
     }
-}
 
+    private void validate(String value) {
+        T parsedValue = fromString.apply(value);
+        if (!validator.test(parsedValue)) {
+            throw DeltaErrors.illegalPropertyValueException(key, value, helpMessage);
```
Comment (dhruvarya-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java:141]: We should try to keep the terminology consistent here: `unknownConfigurationKeyException` and `illegalPropertyValueException`. It should be either Configuration or Property.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java (hunk 9 above)
Comment (EstherBear) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java:159]: Because we may try `doCommit` for multiple times. We only need to update metadata when it's updated during resolving the conflicts.

File: kernel/kernel-api/src/main/java/io/delta/kernel/exceptions/IllegalConfigurationValueException.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+import io.delta.kernel.annotation.Evolving;
+
+/**
+ * Thrown when an illegal value is specified for a table property.
+ *
+ * @since 3.2.0
+ */
+@Evolving
+public class IllegalConfigurationValueException extends KernelException {
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/exceptions/IllegalConfigurationValueException.java:26]: nit: name in `InvalidConfigurationValueException`.

File: kernel/kernel-api/src/main/java/io/delta/kernel/exceptions/UnknownConfigurationKeyException.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+import io.delta.kernel.annotation.Evolving;
+
+/**
+ * Thrown when an unknown configuration key is specified.
+ *
+ * @since 3.2.0
+ */
+@Evolving
+public class UnknownConfigurationKeyException extends KernelException {
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/exceptions/UnknownConfigurationKeyException.java:26]:
```suggestion
public class UnknownConfigurationException extends KernelException {
```

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 17]
```diff
@@ -166,6 +167,74 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
...
+          tablePath,
+          true, testSchema, Seq.empty, Map(TableConfig.CHECKPOINT_INTERVAL.getKey -> "-1"))
+      }
+      assert(
+        ex2.getMessage.contains(
+          String.format(
+            "Invalid value for table property '%s': '%s'. %s",
+            TableConfig.CHECKPOINT_INTERVAL.getKey, "-1", "needs to be a positive integer.")))
+    }
+  }
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:236]: Can you add a test case where the txn is retried?

- Create tabl2
- Create txn1 with config changes
- Create txn2
- Commit txn2
- Try to commit txn1

**Review by vkorukanti — CHANGES_REQUESTED** [review]

File: kernel/kernel-api/src/main/java/io/delta/kernel/TransactionBuilder.java [hunk 18]
```diff
@@ -69,6 +70,20 @@ TransactionBuilder withTransactionId(
...
+     *                   configure the table. And these properties are stored in the table metadata.
+     *
+     * @return updated {@link TransactionBuilder} instance.
+     *
+     * @since 3.2.0
+     */
+    TransactionBuilder withTableProperties(Engine engine, Map<String, String> properties);
+
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/TransactionBuilder.java:86]: We throw the invalid config exceptions as part of the build. Make sure to add them to javadoc as `@throws`. See the existing example `ConcurrentTransactionException`

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 19]
```diff
@@ -893,6 +978,28 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
     }
   }
 
+  private def getMetadata(engine: Engine, table: Table, version: Long): Metadata = {
```
Comment (dhruvarya-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:981]: [nit] this name is slightly misleading. This is the metadata action from commit version X. Maybe `getMetadataActionFromCommit` is better?

Comment (dhruvarya-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:981]: My preference would be for this function to return an Optional[Metadata] to indicate that this can be None.

**Review by dhruvarya-db — APPROVED** [review]

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java [hunk 20]
```diff
@@ -67,7 +67,8 @@ public class TransactionImpl
     private final Path dataPath;
     private final Path logPath;
     private final Protocol protocol;
-    private final Metadata metadata;
+    private Metadata metadata;
+    private boolean shouldUpdateMetadata;
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java:71]: can we revert these changes? Looks like these aren't getting updated anywhere.

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 21]
```diff
@@ -849,6 +978,29 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
     }
   }
 
+  private def getMetadataActionFromCommit(
+    engine: Engine, table: Table, version: Long): Optional[Metadata] = {
+    val logPath = new Path(table.getPath(engine), "_delta_log")
+    val file = engine
+      .getFileSystemClient
+      .listFrom(FileNames.listingPrefix(logPath, version)).next
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:986]: you can directly construct the file status using `FileNames.deltaFile` method.

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 22]
```diff
@@ -849,6 +978,29 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
...
+    val columnarBatches =
+      engine.getJsonHandler.readJsonFiles(
+        singletonCloseableIterator(file),
+        SingleAction.FULL_SCHEMA,
+        Optional.empty())
+    if (!columnarBatches.hasNext) {
+      return null
+    }
+    val metadataVector = columnarBatches.next().getColumnVector(3)
+    Optional.ofNullable(Metadata.fromColumnVector(metadataVector, 1, engine))
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:996]: what happens if the metadata is not a rowId 1? I think we may want to iterate through the actions and find the first non-null `Metadata` object or null if we exhausted the iterator.

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 23]
```diff
@@ -166,6 +170,131 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
...
 
+  test("create table and set properties") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val txn1 = createTxn(engine, tablePath, isNewTable = true, testSchema, Seq.empty)
+
+      txn1.commit(engine, emptyIterable())
+
+      val ver0Snapshot = table.getSnapshotAsOfVersion(engine, 0).asInstanceOf[SnapshotImpl]
+      assertMetadataProp(ver0Snapshot, TableConfig.CHECKPOINT_INTERVAL, 10)
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:181]: This can be one utility method. Also repeated multiple times in tests

```
def setTablePropAndVerify(tablePath: String, isNewTable:Boolean = true, propKey: String, propValue: String) = {
    val table = Table.forPath(engine, tablePath)
      val txn1 = createTxn(engine, tablePath, isNewTable = true, testSchema, Seq.empty)

      txn1.commit(engine, emptyIterable())

      val ver0Snapshot = table.getSnapshotAsOfVersion(engine, 0).asInstanceOf[SnapshotImpl]
      assertMetadataProp(ver0Snapshot, TableConfig.CHECKPOINT_INTERVAL, 10)
}
```

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 24]
```diff
@@ -166,6 +170,131 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
...
+      // Create and commit txn2
+      appendData(
+        engine,
+        tablePath,
+        data = Seq(Map.empty[String, Literal] -> dataBatches1)
+      )
+      // Try to commit txn1
+      txn1.commit(engine, emptyIterable())
+
+      val ver1Snapshot = table.getSnapshotAsOfVersion(engine, 1).asInstanceOf[SnapshotImpl]
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:240]: why not just fetch the latest snapshot?

**Review by vkorukanti — CHANGES_REQUESTED** [review]

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java (hunk 20 above)
Comment (EstherBear) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java]: It will be updated for the following PRs. But it's fine to keep it `final` here.

File: kernel/kernel-api/src/main/java/io/delta/kernel/TransactionBuilder.java [hunk 25]
```diff
@@ -69,6 +73,23 @@ TransactionBuilder withTransactionId(
...
 
+    /**
+     * Set the table properties for the table. When the table already contains the property with
+     * same key, it gets replaced if it doesn't have the same value.
+     *
+     * @param engine     {@link Engine} instance to use.
+     * @param properties The table properties to set. These are key-value pairs that can be used to
+     *                   configure the table. And these properties are stored in the table metadata.
+     * @throws InvalidConfigurationValueException if the value of the property is invalid.
+     * @throws UnknownConfigurationException if any of the properties are unknown to
```
[the remark this answers is not in the document]
Comment (EstherBear) [reply kernel/kernel-api/src/main/java/io/delta/kernel/TransactionBuilder.java]: `build()` Thanks!

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala (hunk 21 above)
Comment (EstherBear) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:963]: Seems `FileNames.deltaFile` method can only return the String?

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala (hunk 22 above)
Comment (EstherBear) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala]: Yes.

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala (hunk 24 above)
Comment (EstherBear) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala]: Yes. I just thought it's clearer to directly fetch a snapshot as of specific version. But it's ok to just fetch the latest one.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java [hunk 26]
```diff
@@ -86,13 +102,52 @@ private TableConfig(
...
+    public static Map<String, String> validateProperties(Map<String, String> configurations) {
+        Map<String, String> validatedConfigurations = new HashMap<>();
+        for (Map.Entry<String, String> kv : configurations.entrySet()) {
+            String key = kv.getKey().toLowerCase(Locale.ROOT);
+            String value = kv.getValue();
+            if (key.startsWith("delta.")) {
+                TableConfig tableConfig = validProperties.get(key);
+                if (tableConfig != null) {
+                    tableConfig.validate(value);
+                    validatedConfigurations.put(tableConfig.getKey(), value);
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableConfig.java:136]: nit: add a comment saying that we want to store the case of the property name defined in TableConfig.

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 27]
```diff
@@ -166,6 +170,108 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
...
+          isNewTable = true,
+          testSchema, Seq.empty, Map(TableConfig.CHECKPOINT_INTERVAL.getKey -> "-1"))
+      }
+      assert(
+        ex2.getMessage.contains(
+          String.format(
+            "Invalid value for table property '%s': '%s'. %s",
+            TableConfig.CHECKPOINT_INTERVAL.getKey, "-1", "needs to be a positive integer.")))
+    }
+  }
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:273]: can you add a test verifying that the case of the property is same as the one in `TableConfig` and not the one passed by the user.

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala [hunk 28]
```diff
@@ -849,6 +955,49 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
...
+        return Optional.ofNullable(Metadata.fromColumnVector(metadataVector, i, engine))
+      }
+    }
+    Optional.empty()
+  }
+
+  def assertMetadataProp(
+    snapshot: SnapshotImpl, key: TableConfig[_ <: Any], expectedValue: Any): Unit = {
+    assert(key.fromMetadata(snapshot.getMetadata) == expectedValue)
+  }
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala:982]: Move these utility methods into `DeltaTableWritesSuiteBase.scala`

**Review by vkorukanti — APPROVED** [review]
Minor comments. Once addressed, this can be merged.

**Review by vkorukanti — APPROVED** [review]
